File: python/traversal_test.py
import unittest

import traversal
from node import Node

"""
    0
  3   2
5       4
"""


class Traversal(unittest.TestCase):
    def setUp(self) -> None:
        self.root = Node(0)
        inner_right = self.root.insert_right(Node(2))
        inner_right.insert_right(Node(4))
        inner_left = self.root.insert_left(Node(3))
        inner_left.insert_left(Node(5))

    def test_inorder__values_should_match_the_order(self):
        result = traversal.inorder_traversal(self.root)
        self.assertEqual(result, [5, 3, 0, 2, 4])

    def test_preorder__values_should_match_the_order(self):
        result = traversal.preorder_traversal(self.root)
        self.assertEqual(result, [0, 3, 5, 2, 4])

    def test_postorder__values_should_match_the_order(self):
        result = traversal.postorder_traversal(self.root)
        self.assertEqual(result, [5, 3, 4, 2, 0])


if __name__ == '__main__':
    unittest.main()

File: c/tests/traversal.test.cpp


#include <catch2/catch.hpp>
#include <node.h>
#include <traversal.h>

TEST_CASE("traversal")
{
    /*
     *      0
     *    3   2
     *   5     4
     *
     * */

    node_s* root = node_create(0);
    node_s* innerRight = node_insert_right(root, 2);
    node_s* innerLeft = node_insert_left(root, 3);
    node_insert_right(innerRight, 4);
    node_insert_left(innerLeft, 5);

    clist_s* list = clist_init(16);

    SECTION("inorder should return the correct order")
    {

        inorder_traversal(root, list);
        int expected[] = { 5, 3, 0, 2, 4 };
        for (int i = 0; i < list->size; ++i)
        {
            REQUIRE(expected[i] == list->data[i]);
        }
    }


    SECTION("preorder should return the correct order")
    {
        preorder_traversal(root, list);
        int expected[] = { 0, 3, 5, 2, 4 };
        for (int i = 0; i < list->size; ++i)
        {
            REQUIRE(expected[i] == list->data[i]);
        }
    }

    SECTION("postorder should return the correct order")
    {
        postorder_traversal(root, list);
        int expected[] = { 5, 3, 4, 2, 0 };
        for (int i = 0; i < list->size; ++i)
        {
            REQUIRE(expected[i] == list->data[i]);
        }
    }
    clist_free(&list);
}
